### PR TITLE
[MacPlatform] Rework Mac toolbar to use a single NSView

### PR DESCRIPTION
**Pull request by iainx**
The Mac toolbar API makes it very hard to make a complex toolbar like we want, creating issues with centring widgets, adding and removing widgets, not updating sizes in real time and enforcing use of the overflow menu. All this leads to very complex and fragile code. Instead of using the NSToolbar API for everything, we create our own toolbar using NSViews and insert this NSView toolbar as the single toolbar item.

### CODE REVIEW

**Comment by iainx** [comment]
@Therzok Can you check I've not missed anything?

File: main/src/addins/MacPlatform/MainToolbar/StatusBar.cs [hunk 1]
```diff
@@ -218,9 +218,9 @@ public StatusBar ()
 
 		public override void DrawRect (CGRect dirtyRect)
 		{
-			if (imageView.Frame.Location == CGPoint.Empty)
+			//if (imageView.Frame.Location == CGPoint.Empty)
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/StatusBar.cs]: Why comment these? They were an optimization to not trigger relayouting.

File: main/src/addins/MacPlatform/MainToolbar/SelectorView.cs [hunk 2]
```diff
@@ -344,14 +341,19 @@ void RealignTexts ()
...
+				PathComponentCells [idx].Title = text;
 				PathComponentCells [ConfigurationIdx].Image = ImageService.GetIcon ("project").ToNSImage ();
 				PathComponentCells [RuntimeIdx].Image = ImageService.GetIcon ("device").ToNSImage ();
 
 				RealignTexts ();
 			}
 
+			void OnSizeChanged ()
+			{
+				SelectorView sview = (SelectorView)Superview;
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/SelectorView.cs]: Style: `var sview = ...`

File: main/src/addins/MacPlatform/MainToolbar/AwesomeBar.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+// THE SOFTWARE.
+using System;
+using AppKit;
+using Foundation;
+using CoreGraphics;
+
+namespace MonoDevelop.MacIntegration.MainToolbar
+{
+	public class AwesomeBar : NSView
+    {
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/AwesomeBar.cs]: Style: spaces

File: main/src/addins/MacPlatform/MainToolbar/MainToolbar.cs [hunk 4]
```diff
@@ -244,102 +96,70 @@ void AttachToolbarEvents (SearchBar bar)
...
+				default:
+					throw new NotImplementedException ();
+				}
 			};
 
 			Action<NSNotification> resizeAction = notif => DispatchService.GuiDispatch (() => {
-				// Skip updates with a null Window. Only crashes on Mavericks.
-				// The View gets updated once again when the window resize finishes.
-				if (bar.Window == null)
+				var win = awesomeBar.Window;
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/MainToolbar.cs]: I've been meaning to change this for some time. Can we not register the notification in AddObserver to a specific window?

File: main/src/addins/MacPlatform/MainToolbar/AwesomeBar.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		internal RunButton RunButton { get; set; }
+		internal SelectorView SelectorView { get; set; }
+		internal StatusBar StatusBar { get; set; }
+		internal SearchBar SearchBar { get; set; }
+		internal ButtonBarContainer ButtonBarContainer { get; private set; }
+
+        public AwesomeBar ()
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/AwesomeBar.cs]: Style: spaces

File: main/src/addins/MacPlatform/MainToolbar/ButtonBarContainer.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		List<ButtonBar> buttonBars;
+		internal List<ButtonBar> ButtonBars {
+			get {
+				return buttonBars;
+			}
+
+			set {
+				buttonBars = value;
+				foreach (var bar in buttonBars) {
+					bar.ResizeRequested += (object sender, EventArgs e) => LayoutButtonBars ();
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/ButtonBarContainer.cs]: Memory-leak alert. Clear resize-requested at some point.

Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/ButtonBarContainer.cs]: Correction: Clear the current `buttonBars` resizeRequested before changing its value. Make the LayoutButtonBars() lambda a function so you can `-= MyAwesomeHandler`.

File: main/src/addins/MacPlatform/MainToolbar/MainToolbar.cs [hunk 7]
```diff
@@ -244,102 +96,70 @@ void AttachToolbarEvents (SearchBar bar)
...
-				MinSize = new CGSize (360, 22),
-				MaxSize = new CGSize (360, 22),
+			selectorView.RuntimeChanged += (sender, ea) => {
+				if (RuntimeChanged != null)
+					RuntimeChanged (sender, ea);
+			};
+
+			widget.WillInsertItem = (tool, id, send) => {
+				switch (id) {
+				case StatusBarId:
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/MainToolbar.cs]: Pedantic: Make this an AwesomeBarId and cleanup the unused values at the top of this class.

File: main/src/addins/MacPlatform/MainToolbar/SelectorView.cs [hunk 8]
```diff
@@ -55,47 +55,41 @@ class OverflowInfoEventArgs : EventArgs
 	[Register]
 	class SelectorView : NSButton
 	{
-		public event EventHandler<SizeRequestedEventArgs> ResizeRequested;
-		public event EventHandler<OverflowInfoEventArgs> OverflowInfoRequested;
+		public event EventHandler<EventArgs> SizeChanged;
 		internal const int ConfigurationIdx = 0;
 		internal const int RuntimeIdx = 1;
 
+		internal PathSelectorView RealSelectorView { get; set; }
```
Comment (Therzok) [inline main/src/addins/MacPlatform/MainToolbar/SelectorView.cs]: Pedantic: `private set`

File: main/src/addins/MacPlatform/MainToolbar/MainToolbar.cs (hunk 4 above)
Comment (iainx) [inline main/src/addins/MacPlatform/MainToolbar/MainToolbar.cs]: Yeah, we can.

**Comment by Therzok** [comment]
Followed up in #1142

**Pull request closed without merging**